39062 cluster variables should be exported properly to secondary storage

mathias-vandaele opened the pull request:
This pull request introduces support for cluster-level variables in the RDBMS schema and service layers. The main changes include adding a new `CLUSTER_VARIABLE` table to the database schema, implementing the necessary Java domain, service, and mapper classes to read and write cluster variables, and updating the main service and writer classes to expose cluster variable functionality.

**Database schema changes:**

* Added a new `CLUSTER_VARIABLE` table to the RDBMS schema, including columns for variable ID, name, type, values, preview flag, resource ID, and scope. Also added an index on `RESOURCE_ID` and Oracle-specific handling for large VARCHAR columns.

**Service and domain layer additions:**

* Introduced `ClusterVariableDbQuery` for querying cluster variables, and `ClusterVariableDbReader` for reading cluster variables from the database, including support for filtering, sorting, paging, and access checks. [[1]](diffhunk://#diff-4c83772ebc36537c90d6f8e22d6a561519afe9111e69e07d6ee31fbb9b4d1b08R1-R89) [[2]](diffhunk://#diff-8592766487dfbb6fa0cbe624dd0cd812c14a4e9f31a467a06ab46082a361d41fR1-R90)
* Added `ClusterVariableMapper` interface for MyBatis mapping of cluster variable queries and mutations.
* Defined `ClusterVariableSearchColumn` enum for cluster variable entity column mapping.

**Integration with main service and writer classes:**

* Updated `RdbmsService` and `RdbmsWriter` classes to include and expose cluster variable reader and writer functionality, including constructor and getter updates. [[1]](diffhunk://#diff-c1a576aeb5cf9e81ae3648856c860c9749b0a4a1556381b3fe1e2452f799cb50R13) [[2]](diffhunk://#diff-c1a576aeb5cf9e81ae3648856c860c9749b0a4a1556381b3fe1e2452f799cb50R59) [[3]](diffhunk://#diff-c1a576aeb5cf9e81ae3648856c860c9749b0a4a1556381b3fe1e2452f799cb50R90) [[4]](diffhunk://#diff-c1a576aeb5cf9e81ae3648856c860c9749b0a4a1556381b3fe1e2452f799cb50R121) [[5]](diffhunk://#diff-c1a576aeb5cf9e81ae3648856c860c9749b0a4a1556381b3fe1e2452f799cb50R190-R193) [[6]](diffhunk://#diff-24b0a420399262a06e830e6c925c13c00fb8046f172da4c8d2226313e1fcf99cR13) [[7]](diffhunk://#diff-24b0a420399262a06e830e6c925c13c00fb8046f172da4c8d2226313e1fcf99cR30) [[8]](diffhunk://#diff-24b0a420399262a06e830e6c925c13c00fb8046f172da4c8d2226313e1fcf99cR72) [[9]](diffhunk://#diff-24b0a420399262a06e830e6c925c13c00fb8046f172da4c8d2226313e1fcf99cL105-R109) [[10]](diffhunk://#diff-24b0a420399262a06e830e6c925c13c00fb8046f172da4c8d2226313e1fcf99cR147) [[11]](diffhunk://#diff-24b0a420399262a06e830e6c925c13c00fb8046f172da4c8d2226313e1fcf99cR212-R215)

**Factory updates:**

* Modified `RdbmsWriterFactory` to accept and manage the new `ClusterVariableMapper` for writer instantiation. [[1]](diffhunk://#diff-1e2289824087f4a8e7786a1984ea655d228b0650d349d6a558ce3592337b2abfR13) [[2]](diffhunk://#diff-1e2289824087f4a8e7786a1984ea655d228b0650d349d6a558ce3592337b2abfR53) [[3]](diffhunk://#diff-1e2289824087f4a8e7786a1984ea655d228b0650d349d6a558ce3592337b2abfL72-R75) [[4]](diffhunk://#diff-1e2289824087f4a8e7786a1984ea655d228b0650d349d6a558ce3592337b2abfR95)

**Minor schema formatting:**

* Reformatted index creation statements for readability in the schema changelog file. [[1]](diffhunk://#diff-e66e70df05eeb3cc91d208ef1e7e433e30f92a6a9f30806a2aa1edae2a228b29L443-R449) [[2]](diffhunk://#diff-e66e70df05eeb3cc91d208ef1e7e433e30f92a6a9f30806a2aa1edae2a228b29L706-R709)


## Related issues

closes https://github.com/camunda/camunda/issues/39062

mathias-vandaele committed:
feat: Modify the RDBMS schema to introduce cluster variable table creation

mathias-vandaele committed:
feat: Introduce cluster variable in the camunda exporter

mathias-vandaele committed:
feat: Introduce cluster variable in the camunda db RDBMS exporter

mathias-vandaele committed:
feat: Introduce camunda cluster variable in schema manager

mathias-vandaele committed:
feat: Introduce camunda cluster variable in the camunda search domain

mathias-vandaele committed:
feat: Introduce camunda cluster variable in the search client for look up

mathias-vandaele committed:
feat: Introduce camunda cluster variable in the search client query transformer

mathias-vandaele committed:
feat: Introduce camunda cluster variable in the camunda search client reader

mathias-vandaele committed:
feat: Introduce configuration for cluster variable in zeebe main

mathias-vandaele committed:
feat: Introduce cluster variable in the RDBMS exporter

mathias-vandaele committed:
feat: Introduce cluster variable in the web app schema

new entity and index for cluster variables

mathias-vandaele committed:
feat: cluster variable value is always a String in the record

This is harmonization with the implementation

mathias-vandaele committed:
feat: cluster variable is now used in zeebe tests

mathias-vandaele committed:
feat: cluster variable is tested in acceptance tests

mathias-vandaele committed:
feat: small fixes after my own review

romansmirnov commented on zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/ClusterVariableHandler.java:
```diff
@@ -0,0 +1,110 @@
+
+  @Override
+  public boolean handlesRecord(final Record<ClusterVariableRecordValue> record) {
+    return SUPPORTED_INTENTS.contains(record.getIntent());
```
```suggestion
    return getHandledValueType().equals(record.getValueType())
        && SUPPORTED_INTENTS.contains(record.getIntent());
```

🔧 Let's make it "safer" - to be fair, not all handlers check for the value type; however, I still recommend adding this check here.

romansmirnov commented on zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/ClusterVariableHandler.java:
```diff
@@ -0,0 +1,110 @@
+      default ->
+          throw new IllegalArgumentException(
+              "Cluster variable with unspecified scope can not be exported");
+    }
```
❌ The handler should avoid throwing an exception; otherwise, the exporter gets stuck, and nothing gets exported anymore at this stage. Instead, there are two options:
1. Assuming that it cannot happen, there is another case than the two; it could log a warn statement like [here](https://github.com/camunda/camunda/blob/cf7513dfa28cf878d67fa3a565f82997262604f0/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/batchoperation/BatchOperationLifecycleManagementHandler.java#L107-L111)
2. Fallback to a default which could be the "global-scoped" cluster variable.

romansmirnov commented on webapps-schema/src/main/java/io/camunda/webapps/schema/entities/clustervariable/ClusterVariableEntity.java:
```diff
@@ -0,0 +1,156 @@
+  private boolean isPreview;
+  private String scope;
+  private String resourceId;
+  @JsonIgnore private Object[] sortValues;
```
```suggestion
```

:x: This should not be necessary; it can be deleted.

romansmirnov commented on webapps-schema/src/main/java/io/camunda/webapps/schema/entities/clustervariable/ClusterVariableEntity.java:
```diff
@@ -0,0 +1,156 @@
+  private String value;
+  private String fullValue;
+  private boolean isPreview;
+  private String scope;
```
🔧 I understand that it could be a String, but to be consistent I would suggest making it out of it an enum to avoid issues.

romansmirnov commented on webapps-schema/src/main/resources/schema/elasticsearch/create/index/camunda-cluster-variable.json:
```diff
@@ -0,0 +1,30 @@
+      "scope": {
+        "type": "keyword"
+      },
+      "resourceId": {
```
❌ Please ensure that this name is consistent across the code.

romansmirnov commented on webapps-schema/src/main/resources/schema/opensearch/create/index/camunda-cluster-variable.json:
```diff
@@ -0,0 +1,30 @@
+      "scope": {
+        "type": "keyword"
+      },
+      "resourceId": {
```
❌ Please ensure that this name is consistent across the code.

romansmirnov commented on zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/ClusterVariableHandlerTest.java:
```diff
@@ -0,0 +1,240 @@
+        factory.generateRecord(ValueType.CLUSTER_VARIABLE, r -> r.withIntent(intent));
+
+    // when - then
+    assertThat(underTest.handlesRecord(decisionRecord)).isFalse();
```
```suggestion
    final Record<ClusterVariableRecordValue> clusterVariableRecord =
        factory.generateRecord(ValueType.CLUSTER_VARIABLE, r -> r.withIntent(intent));

    // when - then
    assertThat(underTest.handlesRecord(clusterVariableRecord)).isFalse();
```

romansmirnov reviewed (changes requested):
Sharing the first batch review feedback that focuses on the `CamundaExporter` and the ES/OS schema.

romansmirnov commented on zeebe/exporters/rdbms-exporter/src/main/java/io/camunda/exporter/rdbms/handlers/ClusterVariableExportHandler.java:
```diff
@@ -0,0 +1,50 @@
+  @Override
+  public boolean canExport(final Record<ClusterVariableRecordValue> record) {
+    return record.getIntent() == ClusterVariableIntent.CREATED
+        || record.getIntent() == ClusterVariableIntent.DELETED;
```
```suggestion
    return record.getValueType() == ValueType.CLUSTER_VARIABLE
        && (record.getIntent() == ClusterVariableIntent.CREATED
        || record.getIntent() == ClusterVariableIntent.DELETED);
```

romansmirnov commented on db/rdbms-schema/src/main/resources/db/changelog/rdbms-exporter/changesets/8.9.0.xml:
```diff
@@ -786,4 +789,31 @@
+    <createIndex tableName="${prefix}CLUSTER_VARIABLE"
+      indexName="${prefix}IDX_CLUSTER_VARIABLE_RESOURCE_ID">
+      <column name="RESOURCE_ID"/>
+    </createIndex>
```
❓ Should it be a composite index? I guess the access pattern would be mostly by scope and resource id, and not just by resource id.

romansmirnov commented on db/rdbms/src/main/java/io/camunda/db/rdbms/write/domain/ClusterVariableDbModel.java:
```diff
@@ -0,0 +1,154 @@
+    private String getCompositeId() {
+      return switch (scope) {
+        case "GLOBAL" -> name;
+        case "TENANT" -> "%s-%s".formatted(name, resourceId);
```
❌ The same problem as in the CamundaExporter, i.e., user might inject their cluster variable by using that ID pattern.

mathias-vandaele commented on zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/ClusterVariableHandler.java:
```diff
@@ -0,0 +1,110 @@
+      case TENANT ->
+          entity
+              .setScope("TENANT")
+              .setResourceId(recordValue.getTenantId())
```
Solution proposal mention

```
{
"name" : String (required),
"value" : String | JSON (required),
"scope" : Object {
"type" : <"GLOBAL" | "TENANT">, (required),
"resourceId" : < null |tenantId> (required when type not "GLOBAL")
}
}
```

This is only the front facing part, but for internal, would you suggest having "resourceId" or "tenantId" ?

romansmirnov reviewed (changes requested):
The next batch of review feedback focused on the RDBMS Exporter. The topics are the same as with the CamundaExporter:
1. Cluster Variable Injection caused by the ID pattern
2. Consistency `resourceId` vs `tenantId`
3. Renaming of `clusterVariableId` to `id`

romansmirnov commented on search/search-domain/src/main/java/io/camunda/search/filter/ClusterVariableFilter.java:
```diff
@@ -0,0 +1,138 @@
+          .valueUntypedOperations(sourceFilter.valueOperations)
+          .scopeOperations(sourceFilter.scopeOperations)
+          .resourceIdOperations(sourceFilter.resourceIdOperations);
+    }
```
```suggestion
```

❌ Not required.

romansmirnov reviewed (changes requested):
Next batch focused on the search-client and search-domain.

romansmirnov commented on search/search-client-query-transformer/src/main/java/io/camunda/search/clients/transformers/filter/ClusterVariableFilterTransformer.java:
```diff
@@ -0,0 +1,60 @@
+
+  private List<SearchQuery> getVariablesQuery(final List<UntypedOperation> variableFilters) {
+    return variableOperations(ClusterVariableIndex.VALUE, variableFilters);
+  }
```
```suggestion
  private List<SearchQuery> getValuesQuery(final List<UntypedOperation> variableFilters) {
    return variableOperations(ClusterVariableIndex.VALUE, variableFilters);
  }
```

romansmirnov reviewed (changes requested):
Next batch focusing on the ES/OS implementation of the Cluster Variable search query.

romansmirnov commented on db/rdbms/src/main/java/io/camunda/db/rdbms/read/service/ClusterVariableDbReader.java:
```diff
@@ -0,0 +1,90 @@
+  private final ClusterVariableMapper clusterVariableMapper;
+
+  public ClusterVariableDbReader(final ClusterVariableMapper clusterVariableMapper) {
+    super(ClusterVariableSearchColumn.values());
```
🔧 Strictly, the search result is not sortable by the ID column

romansmirnov commented on db/rdbms/src/main/java/io/camunda/db/rdbms/read/service/ClusterVariableDbReader.java:
```diff
@@ -0,0 +1,90 @@
+            new ClusterVariableQuery(
+                new Builder().resourceIds(tenant).scopes("TENANT").names(name).build(),
+                ClusterVariableSort.of(b -> b),
+                SearchQueryPage.of(b -> b.from(0).size(1))))
```
```suggestion
            ClusterVariableQuery.of(
                b -> b.filter(f -> f.names(name).scopes("TENANT").resourceIds(tenant)).page(p -> p.from(0).size(1)))
```

🔧 A bit more compact.

romansmirnov commented on db/rdbms/src/main/java/io/camunda/db/rdbms/read/service/ClusterVariableDbReader.java:
```diff
@@ -0,0 +1,90 @@
+            new ClusterVariableQuery(
+                new Builder().scopes("GLOBAL").names(name).build(),
+                ClusterVariableSort.of(b -> b),
+                SearchQueryPage.of(b -> b.from(0).size(1))))
```
```suggestion
            ClusterVariableQuery.of(
                b -> b.filter(f -> f.names(name).scopes("GLOBAL")).page(p -> p.from(0).size(1)))
```

romansmirnov reviewed (changes requested):
The next batch, I will finish the rest on Monday.

mathias-vandaele committed:
feat: Correction of the first review from roman

Partial changes of resourceId to TenantId, changes in id generation and created of deletedHandler

mathias-vandaele committed:
feat: Correction after roman's review

First batch of correction + continuation of renaming resourceId to tenantId

mathias-vandaele committed:
feat: Correction after roman's review 2nd batch

Focus on RDBMS small fixes

mathias-vandaele committed:
feat: Correction after roman's review 3rd batch

mathias-vandaele commented on db/rdbms/src/main/java/io/camunda/db/rdbms/read/service/ClusterVariableDbReader.java:
```diff
@@ -0,0 +1,90 @@
+            new ClusterVariableQuery(
+                new Builder().scopes("GLOBAL").names(name).build(),
+                ClusterVariableSort.of(b -> b),
+                SearchQueryPage.of(b -> b.from(0).size(1))))
```
@romansmirnov In this case, could'nt we use also the getById method as in the documentReader ?

mathias-vandaele commented on search/search-client-query-transformer/src/main/java/io/camunda/search/clients/reader/ClusterVariableDocumentReader.java:
```diff
@@ -0,0 +1,52 @@
+    return getSearchExecutor()
+        .getByQuery(
+            ClusterVariableQuery.of(
+                b -> b.filter(f -> f.names(name).resourceIds(tenant)).singleResult()),
```
@romansmirnov I am currently having `getSearchExecutor().getById(...)` But I don't get how to configure the resource access here ? 

I found other example such as 

```
  @Override
  public AuthorizationEntity getByKey(
      final long key, final ResourceAccessChecks resourceAccessChecks) {
    return getSearchExecutor()
        .getById(
            String.valueOf(key),
            io.camunda.webapps.schema.entities.usermanagement.AuthorizationEntity.class,
            indexDescriptor.getFullQualifiedName());
  }
  ```
  
  I am wondering, how do we know check if the caller has rights to see this resource ?

mathias-vandaele replied:
I think I got it, the `resourceAccessCheck` purely for filtering when making a search (I guess for retrieving less content) because the securityContext is applied anyway at an upper level in this function 

```
 protected <T> T doGetWithResourceAccessController(
      final Function<ResourceAccessChecks, T> applier) {
    return resourceAccessController.doGet(securityContext, applier);
  }
 ```

for single entity, this is not necessary as we will always retrieve a single entity, is it really necessary in that case to add `ResourceAccessCheck` in `getTenantScopedClusterVariable ` and `getGlobalScopedClusterVariable` ?

romansmirnov commented:
@mathias-vandaele, I finished the first round of review. Please let me know, when the feedback is applied and I can re-review 😄

romansmirnov commented on db/rdbms/src/main/java/io/camunda/db/rdbms/write/service/ClusterVariableWriter.java:
```diff
@@ -0,0 +1,50 @@
+            "io.camunda.db.rdbms.sql.ClusterVariableMapper.insert",
+            clusterVariable.truncateValue(
+                vendorDatabaseProperties.variableValuePreviewSize(),
+                vendorDatabaseProperties.charColumnMaxBytes())));
```
Yes, I know that this is the same as the basic variables. I think, this could probably be solved differently. However, it is acceptable to maintain consistency with the basic variables as they are. If needed, this could be refactored in the future.

mathias-vandaele committed:
feat: Correction after roman's review 4th batch

mathias-vandaele committed:
feat: Correction after roman's review 5th batch

mathias-vandaele committed:
feat: Correction after roman's review

Remove the UNSPECIFIED field as it should never happen

mathias-vandaele committed:
feat: Pair programming session with Roman

mathias-vandaele committed:
feat: Tests fix for cluster variables

Not needed anymore, delete does not update and create does not override ID

mathias-vandaele committed:
feat: Composite index was preventing table creation

mathias-vandaele committed:
feat: SQL request would fail without this parenthesis

romansmirnov commented on webapps-schema/src/main/java/io/camunda/webapps/schema/entities/clustervariable/ClusterVariableScope.java:
```diff
@@ -0,0 +1,33 @@
+        yield GLOBAL;
+      }
+    };
+  }
```
```suggestion
  public static ClusterVariableScope fromProtocol(final io.camunda.zeebe.protocol.record.value.ClusterVariableScope scope) {
    return switch (scope) {
      case GLOBAL -> GLOBAL;
      case TENANT -> TENANT;
      // This should never happen, as any other scope would be rejected by the processor before
      // mutating the state, as the safety check, but we need a default case for the switch
      // expression
      default -> {
        LOGGER.error(
            "Unknown cluster variable scope received from protocol: {}. Defaulting to GLOBAL.",
            scope);
        yield GLOBAL;
      }
    };
  }
```

💭 To avoid calling `toString()` when trying to map from one enum to another, I would suggest expecting the actual enum as a parameter.

romansmirnov commented on zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/ClusterVariableCreatedHandler.java:
```diff
@@ -0,0 +1,100 @@
+    implements ExportHandler<ClusterVariableEntity, ClusterVariableRecordValue> {
+
+  private static final ClusterVariableIntent SUPPORTED_INTENT = ClusterVariableIntent.CREATED;
+  private static final Logger LOGGER = LoggerFactory.getLogger(ClusterVariableCreatedHandler.class);
```
```suggestion
```

💭 Some cleanup

romansmirnov commented on search/search-client-query-transformer/src/main/java/io/camunda/search/clients/transformers/filter/ClusterVariableFilterTransformer.java:
```diff
@@ -0,0 +1,84 @@
+    final var matchGlobalQuery =
+        stringTerms(
+            ClusterVariableIndex.SCOPE,
+            List.of(io.camunda.search.entities.ClusterVariableScope.GLOBAL.name()));
```
```suggestion
    final var matchGlobalQuery =
        term(ClusterVariableIndex.SCOPE, ClusterVariableScope.GLOBAL.name());
```

romansmirnov commented on db/rdbms/src/main/java/io/camunda/db/rdbms/read/service/ClusterVariableDbReader.java:
```diff
@@ -0,0 +1,89 @@
+      dbQueryBuilder
+          .tenancyEnabled(true)
+          .authorizedTenantIds(resourceAccessChecks.getAuthorizedTenantIds());
+    }
```
```suggestion
    dbQueryBuilder
        .filter(query.filter())
        .tenancyEnabled(resourceAccessChecks.tenantCheck().enabled())
        .authorizedTenantIds(resourceAccessChecks.getAuthorizedTenantIds())
        .authorizedResourceIds(resourceAccessChecks.getAuthorizedResourceIds())
        .sort(dbSort)
        .page(convertPaging(dbSort, query.page()));
```

💭  Simplifying it.

romansmirnov reviewed (changes requested):
@mathias-vandaele, I have a few findings.

mathias-vandaele committed:
feat: Correction after the 2nd review from Roman

romansmirnov reviewed (approved):
Thanks for applying the review feedback. LGTM.
